src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,135 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] objects;
+    private int size;
+    private int capacity;
+
+    public ArrayList() {
+        this.capacity = DEFAULT_CAPACITY;
+        objects = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= capacity) {
+            capacity = grow();
+        }
+        objects[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (size == capacity) {
+            capacity = grow();
+        }
+        if (index > size() || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index doesn't exist: " + index);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Object[] res = new Object[size + 1];
+        if (index == 0) {
+            res[0] = value;
+        } else {
+            System.arraycopy(objects, 0, res, index - 1, index);
+            res[index] = value;
+        }
+        System.arraycopy(objects, index, res, index + 1, size - index);
+        objects = new Object[capacity];
+        System.arraycopy(res, 0, objects, 0, res.length);
+        size++;
+    }
 
+    private int grow() {
+        capacity += capacity >> 1;
+        Object[] temp = new Object[size];
+        System.arraycopy(objects, 0, temp, 0, size);
+        objects = new Object[capacity];
+        System.arraycopy(temp, 0, objects, 0, size);
+        return capacity;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] arr = new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            arr[i] = list.get(i);
+            add((T) arr[i]);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size() - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "No element with such index in list: " + index);
+        }
```

Comment:
this check is the same as in the set() and remove() methods, let's move it to a private method and reuse